Improve Windows Logo Detection

lazerl0rd opened the pull request:
This also adds support for (the upcoming) Windows 11.

lazerl0rd commented:
The force-pushes were to add Windows 8/8.1 support, alongside fixing the edgecase of a possible "Windows 110" being recognised as "Windows 11".

lazerl0rd commented:
Here's an example of how the Windows 11 ASCII art looks:

![image](https://user-images.githubusercontent.com/24303423/122690907-1cdf8b80-d224-11eb-8d66-1758f190ff14.png)

rashil2000 commented:
I was starting to think our logo detection logic might be getting a little stale after the recent leaks xD

The `switchlogo` parameter is now useless, so it needs to be removed completely (it's still present in your PR).

Also, a more robust way of detecting Windows major version names would be through build numbers - it allows more flexibility and less hard-coded comparisons. So I don't think we should be comparing strings for that. What do you think @jcwillox?

I was about to add Windows 11 logo myself, but it is still not confirmed what build numbers they'll start with (rumor is 22000).

lazerl0rd commented:
> I was starting to think our logo detection logic might be getting a little stale after the recent leaks xD
> 
> The `switchlogo` parameter is now useless, so it needs to be removed completely (it's still present in your PR).
> 
> Also, a more robust way of detecting Windows major version names would be through build numbers - it allows more flexibility and less hard-coded comparisons. So I don't think we should be comparing strings for that. What do you think @jcwillox?
> 
> I was about to add Windows 11 logo myself, but it is still not confirmed what build numbers they'll start with (rumor is 22000).

I've removed the `switchlogo` parameter in a follow-up commit (as shown above).

With regards to using build numbers, we'd then have to keep track (via an array or dictionary) the list of minimum build numbers for each version of Windows (therefore still hard-coding) as opposed to just matching the version string. It isn't much different functionally, but is cleaner to utilise version string matching.

An alternative may be to use the kernel version however the problem with that is that both the "leaked" Windows 11 build and the first set of Windows 10 builds all used kernel versions of their predecessor, and as such would be detected incorrectly.

jcwillox commented:
Haha, might be a little early to be adding Windows 11 support but I'm all for it XD.

I have to agree while build number would seem to be the better method of detection, but matching the version string is probably going to be the easiest and most reliable. Plus any performance difference shouldn't be noticeable.

As for the `-switchlogo` parameter, we should create a new `-logo` parameter to set the logo e.g. `-logo "Windows 11"`.

lazerl0rd commented:
> Haha, might be a little early to be adding Windows 11 support but I'm all for it XD.
> 
> I have to agree while build number would seem to be the better method of detection, but matching the version string is probably going to be the easiest and most reliable. Plus any performance difference shouldn't be noticeable.
> 
> As for the `-switchlogo` parameter, we should create a new `-logo` parameter to set the logo e.g. `-logo "Windows 11"`.

I've added the logo paramater you requested and corrected the previous commits so that they aren't broken if ever ungrouped (hence the force-pushes). Hopefully this meets your criteria.

jcwillox commented on winfetch.ps1:
```diff
@@ -366,46 +365,79 @@ $img = if (-not $noimage) {
+            )
+        } else {
+            Write-Error 'The only version logos supported are Windows 11, Windows 10/8.1/8, and Windows 7/Vista/XP.'
+            exit 1
```
I'm a little unsure if this is the right response, we might be better off just defaulting to the legacy logo if the platform is not Windows 11/10/8? It'd be simpler and safer for unknown versions or edge cases. If you really want to be explicit, you could disable the image and keep going, maybe with a warning as well?

lazerl0rd replied:
@jcwillox, as per Line 374 we do actually default to the legacy logo. This error only appears if you explcitly use the new logo parameter and provide an invalid version of windows (eg. `winfetch.ps1 -o "Windows 95"`)

jcwillox replied:
Oh great, totally missed that

jcwillox reviewed (commented):
Looks pretty good to me

lazerl0rd commented on winfetch.ps1:
```diff
@@ -64,7 +64,7 @@ param(
     [string][alias('c')]$configpath,
     [switch][alias('n')]$noimage,
-    [switch][alias('l')]$switchlogo,
+    [string][alias('o')]$logo,
```
@jcwillox, the reason behind this is because the old switchlogo parameter used an `l` and I assume we want to differentiate between it (just in case anyone is calling Winfetch via a script).

jcwillox replied:
This is going to be a breaking change anyway, and it'll break scripts regardless as PowerShell will either say the flag is missing or assume that `-l` is short for `-logo` and still error out unless you provide a valid OS. At least on PSv7 `-l` and `-o` are treated as the same, so the added `-o` is unnecessary.

rashil2000 replied:
I agree. The `o` is unnecessary. In fact all the single letter aliases are redundant (as of now), because PS allows you to call them as long as they are unambiguous.

lazerl0rd replied:
I've adjusted it back to an `l`, @jcwillox and @rashil2000. My apologies for the delay.

rashil2000 commented:
Great! Thanks for the contribution @lazerl0rd !
